File: shims/common/src/main/java/com/nvidia/spark/rapids/ShimLoader.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.nvidia.spark.rapids.shims;

import java.util.HashMap;
import java.util.Map;

// import com.nvidia.spark.shims.ShimLoaderScala;

/**
 * ShimLoader.
 *
 */
public abstract class ShimLoader {
  public static String SPARK30DATABRICKSSVERSIONNAME = "3.0.0-databricks";
  public static String SPARK30VERSIONNAME = "3.0.0";

  private static SparkShims sparkShims;

  /**
   * The names of the classes for shimming Spark for each major version.
   */
  private static final HashMap<String, String> SPARK_SHIM_CLASSES =
      new HashMap<String, String>();

  static {
    SPARK_SHIM_CLASSES.put(SPARK30VERSIONNAME, "com.nvidia.spark.rapids.shims.Spark300DatabricksShims");
    SPARK_SHIM_CLASSES.put(SPARK30DATABRICKSSVERSIONNAME, "com.nvidia.spark.rapids.shims.Spark300Shims");
  }

  /**
   * Factory method to get an instance of HadoopShims based on the
   * version of Hadoop on the classpath.
   */
  public static synchronized SparkShims getSparkShims() {
    if (sparkShims == null) {
      sparkShims = loadShims(SPARK_SHIM_CLASSES, SparkShims.class);
    }
    return sparkShims;
  }

  private static <T> T loadShims(Map<String, String> classMap, Class<T> xface) {
    String vers = "3.0.0"; // ShimLoaderScala.getVersion();
    String className = classMap.get(vers);
    return createShim(className, xface);
  }

  private static <T> T createShim(String className, Class<T> xface) {
    try {
      Class<?> clazz = Class.forName(className);
      return xface.cast(clazz.newInstance());
    } catch (Exception e) {
      throw new RuntimeException("Could not load shims in class " + className, e);
    }
  }

  /*
  public static String getVersion() {
    String vers = SPARK_VERSION;
    String finalVer = vers;
    // hack for databricks, try to find something more reliable?
    if (SPARK_BUILD_USER.equals("Databricks")) {
        finalVer = vers + "-databricks";
    }
    return finalVer;
  }
  */

  private ShimLoader() {
    // prevent instantiation
  }
}

File: shims/common/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala
/*
 * Copyright (c) 2020, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nvidia.spark.rapids.shims

import scala.collection.immutable.HashMap

import org.apache.spark.{SPARK_BUILD_USER, SPARK_VERSION}

object ShimLoader {

  val SPARK30DATABRICKSSVERSIONNAME = "3.0.0-databricks"
  val SPARK30VERSIONNAME = "3.0.0"

  private var sparkShims: Option[SparkShims] = None

  /**
   * The names of the classes for shimming Spark for each major version.
   */
  private val SPARK_SHIM_CLASSES = HashMap(
    SPARK30VERSIONNAME -> "com.nvidia.spark.rapids.shims.Spark300DatabricksShims",
    SPARK30DATABRICKSSVERSIONNAME -> "com.nvidia.spark.rapids.shims.Spark300Shims"
  )

  /**
   * Factory method to get an instance of HadoopShims based on the
   * version of Hadoop on the classpath.
   */
  def getSparkShims: Unit = {
    if (sparkShims.isEmpty) {
      sparkShims = loadShims(SPARK_SHIM_CLASSES)
    }
    sparkShims
  }

  private def loadShims(classMap: Map[String, String]): SparkShim = {
    val vers = getVersion();
    val className = classMap.get(vers)
    className.map(createShim(_)).orElse(logWarning(s"No shim layer for $vers"))
  }

  private def createShim(className: String): SparkShim = try {
    val clazz = Class.forName(className)
    val res = clazz.newInstance().asInstanceOf[SparkShim]
    res
  } catch {
    case e: Nothing =>
      throw new RuntimeException("Could not load shims in class " + className, e)
  }

  def getVersion(): String = {
    // hack for databricks, try to find something more reliable?
    if (SPARK_BUILD_USER.equals("Databricks")) {
        SPARK_VERSION + "-databricks"
    } else {
      SPARK_VERSION
    }
  }

}
